scalding-core/src/main/scala/com/twitter/scalding/typed/TypedPipe.scala

```diff
@@ -1089,25 +1089,72 @@ final case class MergedTypedPipe[T](left: TypedPipe[T], right: TypedPipe[T]) ext
         case (pipe, cnt) => pipe.flatMap(List.fill(cnt)(_).iterator)
       }
       .map(_.toPipe[U](fieldNames)(flowDef, mode, setter)) // linter:ignore
-      .toList
+      .toVector
 
     if (merged.size == 1) {
       // there is no actual merging here, no need to rename:
       merged.head
     } else {
-      merged.reduce[Pipe] {
-        case (left, right) =>
-          // special handling for cases where one side of the hashjoin is merged
-          // with the hashjoin result. Cascading no longer allows it,
-          // so we add a checkpoint to the join result as a workaround
-          if (RichPipe.isHashJoinedWithPipe(left, right))
-            new Merge(RichPipe.assignName(new Checkpoint(left)), RichPipe.assignName(right))
-          else if (RichPipe.isHashJoinedWithPipe(right, left))
-            new Merge(RichPipe.assignName(left), RichPipe.assignName(new Checkpoint(right)))
-          else
-            new Merge(RichPipe.assignName(left), RichPipe.assignName(right))
-      }
+      val prepared  = mergeAvoidingHashes(merged)
+      prepared
+    }
+  }
+
+  /**
+    * Special handling for cases where one side of the hashjoin is merged
+    *  with the hashjoin result. Cascading no longer allows it (as of 3.0),
+    *  so we insert checkpoints and/or intermediate merge stages as appropriate
+    *
+    * @param pipes a list of pipes possibly colliding with each other
+    * @return an updated list of pipes, which can safely be merged
+    */
+  protected def mergeAvoidingHashes(pipes: Vector[Pipe]): Pipe = {
```

Comment:
Yep, I'll try to tidy this up. I was initially fearing I needed recursion, and aimed for tailrec (which I couldn't achieve). 
Now planning to break ```val safedColliding = match``` as a ```resolveCollidingHalf(colliding)``` method, but unsure how far down that road I should go